### PR TITLE
features: add PlayViewModel and navigation graph

### PR DESCRIPTION
**Pull request by PanVova**

### CODE REVIEW

File: api/src/main/java/com/panhuk/api/di/questionApi/QuestionApiComponent.kt [hunk 1]
```diff
@@ -3,6 +3,7 @@ package com.panhuk.api.di.questionApi
 import com.panhuk.api.api.QuestionApi
 import com.panhuk.api.di.base.BaseApiModule
 import dagger.Component
+import kotlinx.coroutines.CoroutineDispatcher
```
Comment (NickHuk) [inline api/src/main/java/com/panhuk/api/di/questionApi/QuestionApiComponent.kt]: unused dependency

File: build.gradle [hunk 2]
```diff
@@ -16,6 +16,11 @@ buildscript {
   ext.espresso_version = '3.4.0'
   ext.datastore_preferences_version = '1.0.0'
   ext.desugaring_version = '1.1.5'
+  ext.navigation_version = '2.3.5'
+  ext.fragment_ktx_version = '1.4.0'
+  ext.compose_navigation_version = '2.4.0-beta02'
```
Comment (NickHuk) [inline build.gradle]: use only stable libraries versions

File: api/src/main/java/com/panhuk/api/api/QuestionApi.kt [hunk 3]
```diff
@@ -7,10 +7,10 @@ import retrofit2.http.Query
 
 interface QuestionApi {
   @GET("api.php")
-  fun getQuestions(
+  suspend fun getQuestions(
     @Query("amount") amount: Int = 10,
     @Query("category") category: String? = null,
     @Query("difficulty") difficulty: String? = null,
     @Query("type") type: String? = null,
-  ): Flow<BaseResponse>
```
Comment (NickHuk) [inline api/src/main/java/com/panhuk/api/api/QuestionApi.kt:15]: remove import

File: api/src/main/java/com/panhuk/api/api/SessionTokenApi.kt [hunk 4]
```diff
@@ -7,7 +7,7 @@ import retrofit2.http.Query
 
 interface SessionTokenApi {
   @GET("api_token.php")
-  fun getSessionToken(
+  suspend fun getSessionToken(
     @Query("command") command: String = "request"
-  ): Flow<SessionTokenResponse>
```
Comment (NickHuk) [inline api/src/main/java/com/panhuk/api/api/SessionTokenApi.kt:12]: remove import

File: api/src/main/java/com/panhuk/api/di/base/BaseApiModule.kt [hunk 5]
```diff
@@ -4,20 +4,27 @@ import com.panhuk.api.BuildConfig
...
+  fun provideOkhttpClient(loggingInterceptor: HttpLoggingInterceptor): OkHttpClient =
     OkHttpClient.Builder()
+      .addInterceptor(loggingInterceptor)
       .build()
 
   @Provides
   fun provideConverterFactory(): Converter.Factory = GsonConverterFactory.create()
 
+  @Provides
+  fun createLoggingInterceptor(): HttpLoggingInterceptor {
```
Comment (NickHuk) [inline api/src/main/java/com/panhuk/api/di/base/BaseApiModule.kt]: not `create` but `provide` follow the same style for naming

File: menuFeature/src/main/java/com/example/menufeature/MenuFragment.kt [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
+import com.example.menufeature.R
+import com.example.menufeature.databinding.MenuFragmentBinding
+
+class MainActivity : AppCompatActivity() {
```
Comment (NickHuk) [inline menuFeature/src/main/java/com/example/menufeature/MenuFragment.kt]: Why did you rename `MainActivity.kt` to `MainFragment.kt`

File: menuFeature/src/main/java/com/example/menufeature/MenuNavigator.kt [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package com.example.menufeature
```
Comment (NickHuk) [inline menuFeature/src/main/java/com/example/menufeature/MenuNavigator.kt]: package name is terrible. Be more careful, review yourself before requesting other person review

File: playFeature/build.gradle [hunk 8]
```diff
@@ -42,9 +43,16 @@ dependencies {
...
   implementation junit_test.values()
   implementation compose.values()
+  implementation dagger2.values()
+  kapt dagger2_kapt.values()
+  implementation navigation.values()
 
   implementation project(':domain')
   implementation project(':core')
+  implementation project(':api')
+  implementation project(':dataSourceDI')
```
Comment (NickHuk) [inline playFeature/build.gradle]: redundant dependencies, moreover you mixed architectural layers, read some articles about clean architecture or Robert Martin "Clean architecture" book

File: apiDI/src/main/java/com/panhuk/apidi/base/BaseApiModule.kt [hunk 9]
```diff
@@ -4,20 +4,27 @@ import com.panhuk.api.BuildConfig
...
+  fun provideOkhttpClient(loggingInterceptor: HttpLoggingInterceptor): OkHttpClient =
     OkHttpClient.Builder()
+      .addInterceptor(loggingInterceptor)
       .build()
 
   @Provides
   fun provideConverterFactory(): Converter.Factory = GsonConverterFactory.create()
 
+  @Provides
+  fun createLoggingInterceptor(): HttpLoggingInterceptor {
```
Comment (NickHuk) [inline apiDI/src/main/java/com/panhuk/apidi/base/BaseApiModule.kt]: not `create` but `provide`, follow the same style

File: playFeature/build.gradle [hunk 10]
```diff
@@ -42,9 +43,16 @@ dependencies {
...
   implementation junit_test.values()
   implementation compose.values()
+  implementation dagger2.values()
+  kapt dagger2_kapt.values()
+  implementation navigation.values()
 
   implementation project(':domain')
   implementation project(':core')
+  implementation project(':apiDI')
+  implementation project(':dataSourceDI')
```
Comment (NickHuk) [inline playFeature/build.gradle:51]: redundant dependencies, moreover you mixed arch layers, brrrrr.... Read few more articles about clean arch or Robert Martin's book

File: playFeature/src/main/java/com/example/playfeature/PlayFragment.kt [hunk 11]
```diff
@@ -25,16 +27,25 @@ import androidx.compose.ui.tooling.preview.Preview
...
 class PlayFragment : Fragment() {
   private var _binding: PlayFragmentBinding? = null
   private val binding get() = _binding!!
 
+  @Inject
+  protected lateinit var viewModel: PlayViewModel
+
+  override fun onAttach(context: Context) {
+    super.onAttach(context)
+    PlayComponent.create(requireContext().applicationContext).inject(this)
```
Comment (NickHuk) [inline playFeature/src/main/java/com/example/playfeature/PlayFragment.kt:43]: call `inject()` before `super()`

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+      } finally {
+        getQuestions()
+        loadQuestion()
+      }
+    }
+  }
+
+  private suspend fun getCurrentSessionToken() {
+    sessionTokenRepoReader.currentSessionToken.collect { sessionTokenRepo ->
+      sessionToken = sessionTokenRepo!!
```
Comment (NickHuk) [inline playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: Bro, stop doing that `!!`, use kotlin null-safety, add null checks and handle situation when received token is null. This is boner.

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        generateNewSessionToken()
+      } finally {
+        getQuestions()
+        loadQuestion()
+      }
+    }
+  }
+
+  private suspend fun getCurrentSessionToken() {
+    sessionTokenRepoReader.currentSessionToken.collect { sessionTokenRepo ->
```
Comment (NickHuk) [inline playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: Weird name, why `sessionTokenRepo`, `sessionToken` feels better naming

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+      sessionToken = sessionTokenRepo!!
+    }
+  }
+
+  private suspend fun generateNewSessionToken() {
+    sessionTokenRepoReader.generateNewToken().collect { sessionTokenRepo ->
```
Comment (NickHuk) [inline playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: Weird name, why sessionTokenRepo, sessionToken feels better naming

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+  init {
+    viewModelScope.launch(dispatcher) {
+      try {
+        getCurrentSessionToken()
+      } catch (e: Exception) {
+        generateNewSessionToken()
+      } finally {
+        getQuestions()
+        loadQuestion()
```
Comment (NickHuk) [inline playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: this is very weird logic, need to discuss that

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+  private suspend fun generateNewSessionToken() {
+    sessionTokenRepoReader.generateNewToken().collect { sessionTokenRepo ->
+      sessionToken = sessionTokenRepo!!
+    }
+  }
+
+  private suspend fun getQuestions() {
+    getQuestionsUseCase.getQuestions().collect { questionsUseCase ->
+      questions = questionsUseCase!!
```
Comment (NickHuk) [inline playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: Don't use `!!` handle nullability instead

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    }
+  }
+
+  private fun loadQuestion() {
+    viewModelScope.launch {
+      with(questions[numberOfQuestion]) {
+        title = questionTitle
+        questionAnswers = allAnswers
+      }
+      numberOfQuestion++
```
Comment (NickHuk) [inline playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: If you don't plan to allow "view previous question" functionality. Better to use Queue instead of index. Think about that

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  fun checkAnswer(answer: String): Boolean {
+    if (questions[numberOfQuestion].correctAnswer == answer) {
+      loadQuestion()
+      totalScore++
+      return true
+    }
+    loadQuestion()
+    return false
```
Comment (NickHuk) [inline playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: Put your mind to UR code
```
loadQuestion()
return if (questions[numberOfQuestion].correctAnswer == answer) {
      totalScore++
      true
    } else {
        false
    }
```

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  private suspend fun getQuestions() {
+    getQuestionsUseCase.getQuestions().collect { questionsUseCase ->
+      questions = questionsUseCase!!
+    }
+  }
+
+  private fun loadQuestion() {
+    viewModelScope.launch {
```
Comment (NickHuk) [inline playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: why do you start coroutine

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+      with(questions[numberOfQuestion]) {
```
Comment (NickHuk) [inline playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: looks lire `.run` is better solution than `with(){}`

File: settings.gradle [hunk 21]
```diff
@@ -23,3 +23,5 @@ include ':playFeature'
 include ':settingsFeature'
 include ':leaderboardFeature'
```
Comment (NickHuk) [inline settings.gradle:24]: Why do you need that dependencies

File: settings.gradle [hunk 22]
```diff
@@ -23,3 +23,5 @@ include ':playFeature'
...
 include ':apiDI'
+include ':menuFeature'
```
Comment (NickHuk) [inline settings.gradle:26]: Why do you need that dependencies

File: usecaseImpl/src/main/java/com/panhuk/usecaseImpl/GetQuestionsUseCaseImpl.kt [hunk 23]
```diff
@@ -9,9 +9,9 @@ import kotlinx.coroutines.flow.map
 class GetQuestionsUseCaseImpl(private val questionRepository: QuestionRepoReader) :
   GetQuestionsUseCase {
   override fun getQuestions(): Flow<List<Question>?> {
-    return questionRepository.questions.map {
-      it?.map {
-        it.copy(allAnswers = it.allAnswers.shuffled())
+    return questionRepository.questions.map { questions ->
+      questions?.map { question ->
+        question.copy(allAnswers = question.allAnswers.shuffled())
```
Comment (NickHuk) [inline usecaseImpl/src/main/java/com/panhuk/usecaseImpl/GetQuestionsUseCaseImpl.kt]: use `=` instead of `{ return ... }`

**Review by NickHuk — CHANGES_REQUESTED** [review]

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt (hunk 17 above)
Comment (NickHuk) [reply playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: Hash set

File: api/src/main/java/com/panhuk/api/di/questionApi/QuestionApiComponent.kt (hunk 1 above)
Comment (PanVova) [reply api/src/main/java/com/panhuk/api/di/questionApi/QuestionApiComponent.kt]: +

File: build.gradle (hunk 2 above)
Comment (PanVova) [reply build.gradle]: It doesn't work with stable version - so I suppose compose navigation was added very recently

File: api/src/main/java/com/panhuk/api/api/QuestionApi.kt (hunk 3 above)
Comment (PanVova) [reply api/src/main/java/com/panhuk/api/api/QuestionApi.kt:15]: +

File: api/src/main/java/com/panhuk/api/api/SessionTokenApi.kt (hunk 4 above)
Comment (PanVova) [reply api/src/main/java/com/panhuk/api/api/SessionTokenApi.kt:12]: +

File: api/src/main/java/com/panhuk/api/di/base/BaseApiModule.kt (hunk 5 above)
Comment (PanVova) [reply api/src/main/java/com/panhuk/api/di/base/BaseApiModule.kt]: +

File: menuFeature/src/main/java/com/example/menufeature/MenuFragment.kt (hunk 6 above)
Comment (PanVova) [reply menuFeature/src/main/java/com/example/menufeature/MenuFragment.kt]: fixed

File: menuFeature/src/main/java/com/example/menufeature/MenuNavigator.kt (hunk 7 above)
Comment (PanVova) [reply menuFeature/src/main/java/com/example/menufeature/MenuNavigator.kt]: +

File: apiDI/src/main/java/com/panhuk/apidi/base/BaseApiModule.kt (hunk 9 above)
Comment (PanVova) [reply apiDI/src/main/java/com/panhuk/apidi/base/BaseApiModule.kt]: dublicate probably because of merge conflict

File: playFeature/build.gradle (hunk 10 above)
Comment (PanVova) [reply playFeature/build.gradle:51]: dublicate probably because of merge conflict

File: playFeature/src/main/java/com/example/playfeature/PlayFragment.kt (hunk 11 above)
Comment (PanVova) [reply playFeature/src/main/java/com/example/playfeature/PlayFragment.kt:43]: +

File: playFeature/build.gradle (hunk 8 above)
Comment (PanVova) [reply playFeature/build.gradle]: I checked this di implementation is necessary api renamed to -> apiDI

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt (hunk 12 above)
Comment (PanVova) [reply playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: +

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt (hunk 13 above)
Comment (PanVova) [reply playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: +

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt (hunk 14 above)
Comment (PanVova) [reply playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: +

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt (hunk 16 above)
Comment (PanVova) [reply playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: +

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt (hunk 18 above)
Comment (PanVova) [reply playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: +

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt (hunk 19 above)
Comment (PanVova) [reply playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: deleted

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt (hunk 20 above)
Comment (PanVova) [reply playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: +

File: usecaseImpl/src/main/java/com/panhuk/usecaseImpl/GetQuestionsUseCaseImpl.kt (hunk 23 above)
Comment (PanVova) [reply usecaseImpl/src/main/java/com/panhuk/usecaseImpl/GetQuestionsUseCaseImpl.kt]: +

File: playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt (hunk 17 above)
Comment (PanVova) [reply playFeature/src/main/java/com/example/playfeature/PlayViewModel.kt]: I changed that to queue

File: apiDI/src/main/java/com/panhuk/apidi/base/BaseApiModule.kt [hunk 24]
```diff
@@ -4,20 +4,27 @@ import com.panhuk.api.BuildConfig
...
+import okhttp3.logging.HttpLoggingInterceptor
 import retrofit2.Converter
 import retrofit2.Retrofit
 import retrofit2.converter.gson.GsonConverterFactory
 
 @Module
 class BaseApiModule {
   @Provides
-  fun provideOkhttpClient(): OkHttpClient =
+  fun provideOkhttpClient(loggingInterceptor: HttpLoggingInterceptor): OkHttpClient =
```
Comment (NickHuk) [inline apiDI/src/main/java/com/panhuk/apidi/base/BaseApiModule.kt:15]: well done!

File: core/src/main/java/com.panhuk.core/Utils.kt [hunk 25]
```diff
@@ -0,0 +1,3 @@
+package com.panhuk.core
+
+const val ERROR = "ERROR"
```
Comment (NickHuk) [inline core/src/main/java/com.panhuk.core/Utils.kt]: Const or Constants is better name for this file

File: playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+  var totalScore: Int = 0
+
+  init {
+    viewModelScope.launch(dispatcher) {
+      try {
+        generateNewSessionToken()
+        getQuestions()
+        loadQuestion()
+      } catch (e: Exception) {
+        Log.e(ERROR, e.toString())
```
Comment (NickHuk) [inline playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt:37]: I meant error dialogs / Toasts or something that users can interact with. But that's fine/ Let's implement dialogs in further PR's

File: playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+      } else {
+        Log.e(ERROR, "sessionToken is null")
+      }
+    }
+  }
+
+  private suspend fun getQuestions() {
+    getQuestionsUseCase.getQuestions().collect { qst ->
+      if (qst != null) {
+        questions.addAll(qst)
```
Comment (NickHuk) [inline playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt:56]: Use assignment instead of `addAll`. As it may lead to duplicating questions bugs

File: playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+        Log.e(ERROR, "questions are null")
+      }
+    }
+  }
+
+  private fun loadQuestion() {
+    questions.element().run {
+      title = questionTitle
+      questionAnswers = allAnswers
+    }
```
Comment (NickHuk) [inline playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt]: use peek?.run{ ... } instead or handle 'NoSuchElementException' in case of empty Queue

File: playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  fun checkAnswer(answer: String): Boolean {
+    val correctAnswer = questions.element().correctAnswer
```
Comment (NickHuk) [inline playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt]: use `peek()`  or handle `NoSuchElementException`

File: playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt [hunk 30]
```diff
@@ -0,0 +1,81 @@
...
+    questions.remove()
```
Comment (NickHuk) [inline playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt]: use `poll()` or handle `NoSuchElementException`

File: usecaseImpl/src/main/java/com/panhuk/usecaseImpl/GetQuestionsUseCaseImpl.kt [hunk 31]
```diff
@@ -6,13 +6,14 @@ import com.panhuk.useCase.GetQuestionsUseCase
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.map
 
+typealias Questions = Flow<List<Question>?>
```
Comment (NickHuk) [inline usecaseImpl/src/main/java/com/panhuk/usecaseImpl/GetQuestionsUseCaseImpl.kt]: I'm not sure if it is required here. As for me it is a little bit confusing

**Review by NickHuk — CHANGES_REQUESTED** [review]

File: playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt (hunk 27 above)
Comment (PanVova) [reply playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt:56]: What does it mean to use assignment ? questions.add() in a loop ?

Comment (NickHuk) [reply playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt:56]: Assigment means `questions = qst`

**Review by NickHuk — CHANGES_REQUESTED** [review]

File: playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt (hunk 27 above)
Comment (PanVova) [reply playFeature/src/main/java/com/panhuk/playfeature/PlayViewModel.kt:56]: There is problem . You can't make '=' with different collections like Queue and List so you can overload operator or clear every time Queue to renew the data (I chose second variant)

**Comment by PanVova** [comment]
There is problem . You can't make '=' with different collections like Queue and List so you can overload operator or clear every time Queue to renew the data (I chose second variant)

PlayViewModel file
